xapian-core/diversify/diversify.cc

```diff
@@ -0,0 +1,195 @@
+/** @file diversify.cc
+ *  @brief Diversification API
+ */
+/* Copyright (C) 2018 Uppinder Chugh
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/diversify.h"
+
+#include "xapian/error.h"
+#include "debuglog.h"
+
+#include <vector>
+#include <limits>
+#include <algorithm>
+#include <cmath>
+
+using namespace Xapian;
+using namespace std;
+
+Diversify::Diversify(unsigned int k_,
+		     double lambda_,
+		     double b_,
+		     double sigma_sqr_)
+    : k(k_), lambda(lambda_), b(b_), sigma_sqr(sigma_sqr_)
+{
+    LOGCALL_CTOR(API, "Diversify", k_ | lambda_ | b_ | sigma_sqr_);
+}
+
+string
+Diversify::get_description() const
+{
+    return "Diversify()";
+}
+
+void
+Diversify::initialise_points(const MSet &source)
+{
+    LOGCALL_VOID(API, "Diversify::initialise_points", source);
+    TermListGroup tlg(source);
+    for (MSetIterator it = source.begin(); it != source.end(); ++it) {
+	points.push_back(Point(tlg, it.get_document()));
+	weights[it.get_document().get_docid()] = it.get_weight();
```

Comment:
`*it` gives you the docid directly (and without creating an unnecessary temporary `Document` object).

Reply:
Will create modifications for this soon, thanks for pointing it out.